Update foreach APIs to use scalar lists

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52450 wip
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #52449 Add type promotion to a fast path for foreach APIs
* #52448 Support broadcasting via slow path
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51060 Refactor foreach pointwise ops tests to use OpInfo
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* #51058 Refactor foreach binary ops tests with scalars to use OpInfo
* #49712 Refactor foreach unary ops tests to use OpInfo
* #49250 Update foreach binary ops with a single scalar and list of tensors
* #49249 Update foreach binary ops with a scalar list
* #51131 Refactor ForeachUtils.h
* #51894 Refactor ForeachUnaryOp.cu
* **#51893 Update foreach APIs to use scalar lists**

Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

------------------------------------------
Motivation
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef

Testing
Update the tests assuming that any scalar type can be passed now, not just double.

izdeby committed:
Update foreach APIs to use scalar lists

[ghstack-poisoned]

gchanan commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -61,18 +61,18 @@
-    ("aten::_foreach_div", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_div_", datetime.date(2021, 2, 25)),
-    ("aten::_foreach_addcdiv", datetime.date(2021, 2, 25)),
-    ("aten::mkldnn_linear", datetime.date(2021, 3, 2)),
```
what is happening with mkldnn_linear here?

gchanan commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -128,7 +119,7 @@ bool can_use_fast_route(TensorList tensors, Scalar scalar) {
     }
 
-    if (!will_promote_tensor(t, scalar)) {
+    if (will_promote_tensor(t, scalar)) {
```
wait, was this computing the wrong thing before?

izdeby replied:
no, the name was wrong

iseeyuan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6749,49 +6749,49 @@
     CUDA: foreach_tensor_div_list_kernel_cuda_
 
-- func: _foreach_add.ScalarList(Tensor[] tensors, float[] scalars) -> Tensor[]
+- func: _foreach_add.ScalarList(Tensor[] tensors, Scalar[] scalars) -> Tensor[]
```
I don't think it breaks backward compatibility. Could you confirm? What would happen if a scalar other than float is passed into an old model, in the new code, unlikely?

izdeby replied:
no BC because those are private unused APIs.

Commit:
Update on "[wip] Update foreach APIs to use scalar lists"


Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

[ghstack-poisoned]

Commit:
Update on "[wip] Update foreach APIs to use scalar lists"


Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

[ghstack-poisoned]

izdeby commented:
@ngimel, fixed the build! This is the [fix](https://github.com/pytorch/pytorch/pull/51893/commits/c70832e4444afdba3a8183820cc554a062813e1c). The problem was was incorrect check for strides which caused some cases to go via fast path when it should have gone via slow.

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

------------------------------------------
Motivation
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef

Testing
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

------------------------------------------
Motivation
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef

Testing
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

------------------------------------------
Motivation
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef

Testing
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

------------------------------------------
Motivation
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef

Testing
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

------------------------------------------
Motivation
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef

Testing
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D26323606](https://our.internmc.facebook.com/intern/diff/D26323606)

------------------------------------------
Motivation
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef

Testing
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]
